# **APA - Data Preparation Template**
## **DATA 3300**

## **Name(s):**

### **Read and follow these assignment instructions carefully! Ordinarily, we’d jump into cleansing data as needed and we’d each do it slightly differently. This is difficult to grade, so please go in the order of this document and follow these instructions.**

**For this assignment students should submit one .ipynb file, one clean, sorted Excel file, and a PDF of this iPython notebook file. See the assignment background for attribute descriptions and the assignment requirements link on the Canvas assignment page.**

**Students should alter code in lines with "replace code..." language. Otherwise, students do not need to alter the provided code.**

## **Q1**

**What does the term “data quality” refer to and why is it important within the context of using data to solve business problems?**

### Data quality refers to how accurate and "clean" data is. It's important because with unlcean data, misrepresantation is present. Solving business problems with poor data will lead to innacurate results.

##**Q2**

**The lecture and textbook discussed five characteristics of data quality: accuracy, completeness, consistency, timeliness, and uniqueness. While “accuracy” in this dataset is hard for us to gauge without further inquiry and“timeliness” kind of depends on what it is the data are to be used for, there are clear examples of problems with each of the other three characteristics.**

**For each of the other three (completeness, consistency, uniqueness), identify a specific situation within the *hbdata-orig.csv* dataset (identify record numbers when applicable). Hint: For uniqueness, look at the VisitSpan attribute. Be sure your answers are clearly labelled and described.**

**This can be done by viewing the data in Excel or importing it into Python. If using Python, begin by importing your libraries and then read in the .csv file.**

In [1]:
import pandas as pd
import numpy as np

#just run this code block after importing libraries above, no need to alter the below code!
from datetime import datetime #you'll also need this method to convert a variable to date-time later in the assignment
import warnings
warnings.filterwarnings("ignore") #this can be removed but will help the code be a little cleaner by ignoring warnings

In [2]:
#replace with code for importing the dataset
df = pd.read_excel('/content/hbdata-orig24.xlsx')
#run this code block after importing dataset above
pd.set_option("display.max_rows", None, "display.max_columns", None) #View full dataframe
df

VisitID             BirdName  BirdType BirdSize  Gender  \
0       2203             Rose Kim     Rufus        S  female   
1       2204       Horace Prefect  Calliope        M    male   
2       2205   Edeltraud McKnight    Rufous        M  female   
3       2206     Herbert McKnight    Rufous        M    male   
4       2207    Walpurga Schwalbe    Anna's        M  female   
5       2208    Walpurga Schwalbe    Anna's        M  female   
6       2209                  NaN       NaN      NaN     NaN   
7       2210          Blossom Kim    Rufous        M  female   
8       2211          Gale Sharma    Rufous       XL  female   
9       2212        King Schwalbe    Anna's        L    male   
10      2213            Lucky Kim    Rufous        M    male   
11      2214        Ruby Schwalbe    Anna's        S  female   
12      2215          Norm Sharma    Rufous        M    male   
13      2216   Petunia O'Flaherty    Anna's        S  female   
14      2217         Sol Schwalbe    Anna's        M    male   
15      2218         Mitz Johnson     Rufus        M  female   
16      2219     Francine Johnson    Rufous        M  female   
17      2220        June McKnight    Rufous        M  female   
18      2221            Lucky Kim    Rufous        M    male   
19      2222        King Schwalbe    Anna's        L    male   
20      2223      Virgil Schwalbe    Anna's        M    male   
21      2224      Bort O'Flaherty    Anna's        L    male   
22      2225     Malachi Schwalbe    Anna's       XL    male   
23      2226        June McKnight    Rufous        M  female   
24      2227        Bobert Sharma    Rufous        M    male   
25      2228   Petunia O'Flaherty    Anna's        S  female   
26      2229        Ruby Schwalbe    Anna's        S  female   
27      2230   Edeltraud McKnight    Rufous        M  female   
28      2231        King Schwalbe    Anna's        L    male   
29      2232    Walpurga Schwalbe    Anna's        M  female   
30      2233        Renee Prefect  Calliope        M  female   
31      2234        June McKnight    Rufous        M  female   
32      2235   Edeltraud McKnight    Rufous        M  female   
33      2236   Petunia O'Flaherty    Anna's        S  female   
34      2237      Clancy McKnight    Rufous        M    male   
35      2238      Bort O'Flaherty    Anna's        L    male   
36      2239     Francine Johnson       NaN        M  female   
37      2240        Ruby Schwalbe    Anna's        S  female   
38      2241      Bort O'Flaherty    Anna's        L    male   
39      2242            Lucky Kim     Rufus        M    male   
40      2243        King Schwalbe    Anna's        L    male   
41      2244               Horace   Prefect      NaN      no   
42      2245         Mitz Johnson    Rufous        M  female   
43      2246        Renee Prefect  Calliope        M  female   
44      2247         Wally Sharma    Rufous        M    male   
45      2248         Mitz Johnson    Rufous        M  female   
46      2249        Renee Prefect  Calliope        M  female   
47      2250   Edeltraud McKnight    Rufous   medium  female   
48      2251        Renee Prefect  Calliope   medium  female   
49      2252      Judith McKnight    Rufous   medium  female   
50      2253          Blossom Kim    Rufous   medium  female   
51      2254         Cake Prefect  Calliope   medium    male   
52      2255      Virgil Schwalbe       NaN   medium    male   
53      2256       Horace Prefect  Calliope   medium    male   
54      2257      Judith McKnight    Rufous   medium  female   
55      2258     Herbert McKnight    Rufous        M    male   
56      2259          Gale Sharma    Rufous       XL  female   
57      2260     Hortence Prefect  Calliope        M  female   
58      2261         Wally Sharma    Rufous        M    male   
59      2262     Hortence Prefect  Calliope        M  female   
60      2263       Kevin McKnight    Rufous        M    male   
61      2264      Bill O'Flaherty    


*   Completeness: VisitIDs 2209, 2395, missing data
*   Consistency: VisitIDs 2250-2257, medium instead of M. Lots of oz instead of ml as well
*   Uniqueness: 2235-2336: Time is the same, duplicate entry

## **Q3**

**Answer the following questions regarding the importance of each of the named data scrubbing steps.**

### **3A**
**What negative outcome would likely occur during data analysis if duplicated visits were not removed (consolidated)?**

The dulicate data would be recorded multiple times, scewing the results of a particular bird or feeder

### **3B**

**What negative outcome would likely occur during data analysis if Location entries were not standardized?**

Some locations would have missing entries, for example "Deck. 2" would not be counted in "Deck 2"

### **3C**

**What negative outcome would likely occur during data analysis if, due to measurement issues, amount eaten values were not consistent?**

One might assume a bird has eaten much more or much less than it actually has. This can cause future predictions of how much to feed to be wrong.

## **Q4**

**In scrubbing the data, you should address each of the following. It’s recommended to go through this assignment in the order of steps below.**

### 4A

In Python trim all text columns to ensure there are no leading or trailing spaces on any of the data values. If you still see two seemingly identical entries, remember that trimming does not trim internal spaces between two words.

In [3]:
df_obj = df.select_dtypes(['object']) #select all columns containing text (of type object)
df[df_obj.columns] = df_obj.apply(lambda x: x.str.strip()) #trim all selected columns
#replace with code to view the header of the dataframe (cheat sheet!)
df.head()
#keep below code to verify date/time of your work
from datetime import datetime
now = datetime.now()
formatted_time = now.strftime("%Y-%m-%d %H:%M:%S")
print(f"Current Date and Time: {formatted_time}")

Current Date and Time: 2024-01-18 21:31:09


### 4B

Ensure each bird’s first and last name should be included in separate fields (columns).

In [4]:
#replace with code to split column, updating column names and delimiter type (cheat sheet)
df[['BirdFirstName','BirdLastName']] = df['BirdName'].str.split(' ', expand=True)
#replace with code to drop birdname column
df = df.drop('BirdName', axis = 1)
#replace with code to show df heading
df.head()

VisitID  BirdType BirdSize  Gender VisitDate                  VisitSpan  \
0     2203     Rufus        S  female       NaN  23:13:05.736-23:13:21.109   
1     2204  Calliope        M    male       NaN    07:45:07.65-07:45:17.97   
2     2205    Rufous        M  female       NaN    08:09:52.10-08:10:22.74   
3     2206    Rufous        M    male       NaN    08:41:54.07-08:42:07.48   
4     2207    Anna's        M  female       NaN    09:18:59.33-09:19:13.41   

     Location Hoverfeed Amount Unnamed: 10 BirdFirstName BirdLastName  
0  Back Patio       yes  17 ml         NaN          Rose          Kim  
1   NE Corner       yes   6 ml         NaN        Horace      Prefect  
2      Deck 2        no   5 ml         NaN     Edeltraud     McKnight  
3      Deck 1       yes   2 ml         NaN       Herbert     McKnight  
4      Garage       yes   3 ml         NaN      Walpurga     Schwalbe

### 4C
Ensure that data values are standardized in each column (pay particular attention to the textual columns, BirdName, BirdType, BirdSize, Gender, Location, and Hoverfeed) so that each “idea” (e.g., a species of bird) is specified (i.e., spelled) the same way.

This can be checked for each column by using the .value_counts() method! After applying the standarization, show that the correction has been made by using the .value_counts() method to display all categories for each column of interest.

In [5]:
#Replace with code to examine levels in BirdType using the value_counts() method, update column name
df['BirdType'].value_counts()

Rufous      103
Anna's       85
Calliope     61
Rufus        18
Prefect      14
Name: BirdType, dtype: int64

In [6]:
#Replace with code to replace redundant value in BirdType, for only 1 original and new label
df = df.replace('Rufus','Rufous')

In [7]:
#Replace with code to examine levels of BirdSize. Ignore null, "medium" and 20ml lines; these will be fixed later.
df['BirdSize'].value_counts()

M         162
S          48
L          33
XL         23
medium     13
20 ml       5
Name: BirdSize, dtype: int64

In [8]:
#Replace with code to replace redundant value in BirdSize
df = df.replace('medium','M')

In [15]:
#Replace with code to examine levels of Location. Ignore two lines with all numerical output; these will be fixed later.
df['Location'].value_counts()

Back Patio                 76
Deck 2                     61
Deck 1                     46
NE Corner                  44
Garage                     29
Entry                      29
15:16:40.19-15:16:55.98     4
13:14:33.36-13:15:09.25     2
Name: Location, dtype: int64

In [14]:
#Replace with code to replace redundant values of Location, for 2 original and new labels
df = df.replace({'Back  Patio':'Back Patio','Deck. 2':'Deck 2'})

### 4D
The amount eaten is very important to our analysis. Where this value is null (blank), omit the record. Note that values of 0 are not the same as null. Retain (don’t delete) the records where the value of Amount is 0. Other issues with this attribute will be addressed later.

Before removing missing columns first examine the length of the dataset currently using the len() method. Then split the *Amount* column into two columns, *Amount_Eat* and *Unit*. Finally, change *Amount_Eat* to a float object using pd.to_numeric().

Once missing values are dropped, view the length of the dataset again using the len() method to see how many observations were dropped.

In [16]:
len(df) #return length of dataframe (i.e., number of rows), just run this code block

291

In [ ]:
#Replace with code to create new split columns for 'Amount_Eat' and 'Unit' from the 'Amount' column, updating column names and delimiter type
df[['Amount_Eat','Unit']] = df['Amount'].str.split(' ', expand=True)
#Replace with code to drop 'Amount' column
df = df.drop('Amount', axis = 1)
#Replace with code to show header of dataframe
df.head()

In [ ]:
df['Amount_Eat'] = df['Amount_Eat'].apply(pd.to_numeric, errors='coerce') #convert Amount_Eat column variable dtype from object to float
#just run this code block
df

In [20]:
#Replace with code to remove missing data for the 'Amount_Eat' column
df = df[df['Amount_Eat'].notna()] # drop rows with missing data based on a specified column
len(df) #check the new length of the dataset

283

### 4E

Ensure that each visit is unique. In other words, no visit should be recorded more than once (duplicated). Remove duplicates from the VisitSpan column  using the drop_duplicates() method.

Then use the len() method to show how many observations were dropped.

In [21]:
df = df.drop_duplicates(subset=['VisitSpan']) #drop duplicate records from the 'VisitSpan' column

#Replace with code to check the new length of the dataset (see above)
len(df)

279

### 4F
Split the VisitSpan column into two columns: rename them StartTime and EndTime.

Then ensure that both time columns are set as datetime objects by using the pd.to_datetime() method.

In [ ]:
#Replace with code to split 'VisitSpan' based on the '-' delimiter, create a StartTime and EndTime column
df[['StartTime', 'EndTime']] = df['VisitSpan'].str.split('-', expand = True) # split based on _ delimeter
df['StartTime'] = pd.to_datetime(df['StartTime'], format='%H:%M:%S.%f') #converts StartTime to hours:minutes:seconds format
df['EndTime'] = pd.to_datetime(df['EndTime'], format='%H:%M:%S.%f') #converts EndTime to hours:minutes:seconds format

#Replace with code to drop the 'VisitSpan' column
df = df.drop('VisitSpan', axis = 1)
#Replace with code to show header of dataframte
df.head()

### 4G
Some recordings of Amount appear to have been made in ounces rather than in milliliters. All of them should be in milliliters. Convert those in ounces to milliliters (1 oz. = 29.57 ml).

Start by subsetting the data to only those observations containing oz, then apply the conversion to the *Amount_Eat* column and replace the values in the *Unit* column with ml.


In [ ]:
df_oz = df[df['Unit'] == 'oz'] #subsets dataset to only contain oz
df_oz['Amount_Eat'] = 29.57*df['Amount_Eat'] #applies conversion formula
df_oz['Unit'] = df_oz['Unit'].replace('oz', 'ml') #replaces oz with ml
df_oz.head() #just run this entire code block as is

###**4H**

**Next, you'll combine your original dataframe with the new subset dataframe using the concat() method. Finally, drop rows containing 'oz' in the *Unit* column.**

**Make sure you have the correct number of observations still by using the len() method.**

In [24]:
frames = [df, df_oz]
df = pd.concat(frames) #combines the original df with the new subsetted df
df = df[df.Unit != 'oz'] #removes rows containing oz

#Replace with code to examine length of dataframe (see earlier portions of this assignment)
len(df)

279

### 4I

Createa new column, called “IsFemale”. This column should include a 1 if the hummingbird that made the visit is female, or 0 if the hummingbird is male.

In [ ]:
#Replace with code to create new column called IsFemale using the np.where method
df['IsFemale'] = np.where(df['Gender'] == 'female',1,0)
df.head()

### 4J
Create a new column called Duration, which includes the duration of the visit in seconds. Creat this column from the *StartTime* and *EndTime* columns then convert to seconds using dt.total_seconds().

In [ ]:
df['Duration'] = df['EndTime'] - df['StartTime'] #creates new 'Duration' column by subtracting StartTime from EndTime

In [ ]:
df['Duration'] = df['Duration'].dt.total_seconds() #converts 'Duration' to seconds
df.head()

### 4K

View the full dataset, to spot check your cleaned data and address any remaining issues you notice.

In [ ]:
pd.set_option("display.max_rows", None, "display.max_columns", None) #views full dataframe
df

In [ ]:
#Replace with code to remove any remaining unwanted columns
df.head()

In [ ]:
pd.set_option("display.max_rows", None, "display.max_columns", None) #views full dataframe
df

## Q5

When you’re finished cleaning the datafile, sort the file by Visit ID (low to high), then save the cleaned file as an Excel document for submission via Canvas.

In [ ]:
#Replace with code to set index to VisitID
df.sort_index() #sorts index

In [ ]:
#export your cleaned Excel file for submission, along with your notebook

#For Google Colab
from google.colab import files
df.to_excel("cleaned_data_HB.xlsx")
files.download("cleaned_data_HB.xlsx")

#For JupyterNotebooks
df.to_excel("replace_with_file_path/file_name.xlsx")

#keep below code to verify date/time of your work
from datetime import datetime
now = datetime.now()
formatted_time = now.strftime("%Y-%m-%d %H:%M:%S")
print(f"Current Date and Time: {formatted_time}")

For submission to Canvas: Download/Save your .ipynb file, print your .ipynb file to PDF, and export your cleaned dataset as an Excel file.